man/vendstat.8
.TH VENDSTAT 8 "February 10, 2010" "OpenIB" "OpenIB Diagnostics"

.SH NAME
vendstat \- query InfiniBand vendor specific functions

.SH SYNOPSIS
.B vendstat
[\-d(ebug)] [\-G(uid)] [\-N] [\-w] [\-i] [\-c <num,num>]
[\-C ca_name] [\-P ca_port] [\-t(imeout) timeout_ms]
[\-V(ersion)] [\-h(elp)] <lid|guid>

.SH DESCRIPTION
.PP
vendstat uses vendor specific MADs to access beyond the IB spec
vendor specific functionality. Currently, there is support for
Mellanox InfiniSwitch-III (IS3) and InfiniSwitch-IV (IS4).

.SH OPTIONS

.PP
.TP
\fB\-N\fR
show IS3 or IS4 general information.
.TP
\fB\-w\fR
show IS3 port xmit wait counters.
.TP
\fB\-i\fR
show IS4 counter group info.
.TP
\fB\-c\fR <num,num>
configure IS4 counter groups.

Configure IS4 counter groups 0 and 1. Such configuration
is not persistent across IS4 reboot.
First number is for counter group 0 and second is for
counter group 1.

Group 0 counter config values:
 0 - PortXmitDataSL0-7
 1 - PortXmitDataSL8-15
 2 - PortRcvDataSL0-7

Group 1 counter config values:
 1 - PortXmitDataSL8-15
 2 - PortRcvDataSL0-7
 8 - PortRcvDataSL8-15

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-d      raise the IB debugging level.
        May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
        May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Addressing flags
.PP
\-G      use GUID address argument. In most cases, it is the Port GUID.
        Example:
        "0x08f1040023"
.PP
\-s <smlid>      use 'smlid' as the target lid for SM/SA queries.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.

Multiple CA/Multiple Port Support

When no IB device or port is specified, the port to use is selected
by the following criteria:
.PP
1. the first port that is ACTIVE.
.PP
2. if not found, the first port that is UP (physical link up).

If a port and/or CA name is specified, the user request is
attempted to be fulfilled, and will fail if it is not possible.

.SH EXAMPLES

.PP
vendstat -N 6		# read IS3 or IS4 general information
.PP
vendstat -w 6		# read IS3 port xmit wait counters
.PP
vendstat -i 6 12	# read IS4 port 12 counter group info
.PP
vendstat -c 0,1 6 12	# configure IS4 port 12 counter groups for PortXmitDataSL
.PP
vendstat -c 2,8 6 12	# configure IS4 port 12 counter groups for PortRcvDataSL

.SH AUTHOR
.TP
Hal Rosenstock
.RI < hal.rosenstock@gmail.com >


ChangeLog
commit c3cc42fffb2d1266cb2c3389a5fa3a8fd4f3f9e9
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:52:23 2009 -0500

    Spin 1.2.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit bfdca7d4392aaee9d8253d306378b036b8a60874
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:52:22 2009 -0500

    Remove immediate work requests
    
    The aren't supported/defined in iWARP.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 7b7460e47ad82ea81101230642c43ee10c6ad517
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:18:12 2009 -0500

    Handle 0B IO better.
    
    Allow num_sge == 0 to indicate a 0B IO.
    
    Also, for read, if length is 0 then create a valid 0B read request
    regardless of the user's specified lkey/rkey.  This avoids 0 stag
    connection terminations.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit e901d7d2ab0e9429118765e98e69a3992b91c584
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Sep 29 14:00:54 2009 -0500

    Reset the T3 wq when the qp moves to RESET.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0f71e2dc0e1a40f4954f30f14deae965818c3c4e
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:10:57 2009 -0500

    Spin release 1.2.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 76d2a6347ab753d26237dc94c95da8d0c6c0c251
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:07:44 2009 -0500

    Support for doorbell overflow avoidance.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ad77c21cafde669c878032fc0f46635408568498
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Mon Sep 28 14:07:20 2009 -0500

    Support sq_sig_all QP attribute.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f6ea9eb207fa4b68f3a9c6f016bc8a262edb1808
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:58:03 2009 -0500

    Spin release version 1.2.3.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 0ec532c2b6b92be6adf58849c9e3d215a50d306d
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Jul 22 12:52:51 2009 -0500

    Support new device ids: 0x35, 0x36.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit f685c8fe7e77e64614d825e563dd9f02a0b1ae16
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:46 2008 -0500

    Spin release 1.2.2
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ebd19d077d331270805aefa9ca00d3661561ba87
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Jul 17 11:28:27 2008 -0500

    Relax the required firmware level.
    
    Turns out that libcxgb3 can really run on either 6.0 or 5.0
    firmware.  So I'm relaxing the requirement...
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 6f7485feb244d8571fcab2292ef92c97bea48df0
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue May 13 09:09:35 2008 -0500

    Spin release 1.2.1.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 309916b23ba384a0bd297e7e51b6b89bff357a78
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:45:01 2008 -0500

    Use local stag == 1 to silently drop the p2p read response.
    
    The current code fails if there are read completions -and-
    the p2p read response in the cq when its flushed.
    
    This fix uses the fact that the RTR read is to target stag 1,
    which is not a valid stag.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 67d519c4a9e28e54a9e436404156c9ae7eb60314
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed May 7 20:25:41 2008 -0500

    Handle wrapping when flushing the software sq.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit a3dcea751a94ad5e0a9cad6c55a839a9848f327c
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Spin 1.2.0 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit cc2ef741842883233a77e18d65a828c711e581a3
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Sat Apr 26 11:39:43 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit 92322b93f48fa98b4fadb736ae49dafda237272a
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Thu Apr 24 14:23:50 2008 -0500

    Support for peer-to-peer connection setup.
    
    - ignore unsolicited 0B reads.

commit a8fb6221af4d914f39c1e71f9331f6d428d698c2
Author: Steve Wise (Open Grid Computing) <swise@hosting.openfabrics.org>
Date:   Tue Apr 1 15:30:11 2008 -0700

    Spin 1.1.5 release.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit db41ebc10f4cc4c2fe7826589636ab7d4b804218
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 26 08:27:12 2008 -0600

    Provide an empty stub for iwch_query_qp.
    
    This prevents crashes when calling ibv_query_qp due to an expected
    library provided function.
    
    Signed-off-by: Jon Mason <jon@opengridcomputing.com>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit ec4f7f1c6b8beceb9001519bc4e2ee89fbb2e6d0
Author: Dotan Barak <dotanb@dev.mellanox.co.il>
Date:   Tue Feb 26 08:23:36 2008 -0600

    Remove duplicated line.
    
    Signed-off-by: Dotan Barak <dotanb@dev.mellanox.co.il>
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 8218482d03898f859e1fa272649fafc64f91f05f
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Tue Feb 12 15:16:03 2008 -0600

    Spin release 1.1.4.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 1bc6930d8884646ef73db71a95cc32c96c20ba32
Author: Jon Mason <jon@opengridcomputing.com>
Date:   Tue Feb 12 14:46:10 2008 -0600

    libcxgb3: zeroing of wc_flags
    

src/vendstat.c
/*
 * Copyright (c) 2004-2009 Voltaire Inc.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <netinet/in.h>

#include <infiniband/umad.h>
#include <infiniband/mad.h>

#include "ibdiag_common.h"

#define IS3_DEVICE_ID			47396

#define IB_MLX_VENDOR_CLASS		10
/* Vendor specific Attribute IDs */
#define IB_MLX_IS3_GENERAL_INFO		0x17
#define IB_MLX_IS3_CONFIG_SPACE_ACCESS	0x50
#define IB_MLX_IS4_COUNTER_GROUP_INFO   0x90
#define IB_MLX_IS4_CONFIG_COUNTER_GROUP 0x91
/* Config space addresses */
#define IB_MLX_IS3_PORT_XMIT_WAIT	0x10013C

struct ibmad_port *srcport;

typedef struct {
	uint16_t hw_revision;
	uint16_t device_id;
	uint8_t reserved[24];
	uint32_t uptime;
} is3_hw_info_t;

typedef struct {
	uint8_t resv1;
	uint8_t major;
	uint8_t minor;
	uint8_t sub_minor;
	uint32_t build_id;
	uint8_t month;
	uint8_t day;
	uint16_t year;
	uint16_t resv2;
	uint16_t hour;
	uint8_t psid[16];
	uint32_t ini_file_version;
} is3_fw_info_t;

typedef struct {
	uint8_t resv1;
	uint8_t major;
	uint8_t minor;
	uint8_t sub_minor;
	uint8_t resv2[28];
} is3_sw_info_t;

typedef struct {
	uint8_t reserved[8];
	is3_hw_info_t hw_info;
	is3_fw_info_t fw_info;
	is3_sw_info_t sw_info;
} is3_general_info_t;

typedef struct {
	uint32_t address;
	uint32_t data;
	uint32_t mask;
} is3_record_t;

typedef struct {
	uint8_t reserved[8];
	is3_record_t record[18];
} is3_config_space_t;

#define COUNTER_GROUPS_NUM 2

typedef struct {
	uint8_t reserved1[8];
	uint8_t reserved[3];
	uint8_t num_of_counter_groups;
	uint32_t group_masks[COUNTER_GROUPS_NUM];
} is4_counter_group_info_t;

typedef struct {
	uint8_t reserved[3];
	uint8_t group_select;
} is4_group_select_t;

typedef struct {
	uint8_t reserved1[8];
	uint8_t reserved[4];
	is4_group_select_t group_selects[COUNTER_GROUPS_NUM];
} is4_config_counter_groups_t;

void counter_groups_info(ib_portid_t * portid, int port)
{
	char buf[1024];
	ib_vendor_call_t call;
	is4_counter_group_info_t *cg_info;
	int i, num_cg;

	memset(&call, 0, sizeof(call));
	call.mgmt_class = IB_MLX_VENDOR_CLASS;
	call.method = IB_MAD_METHOD_GET;
	call.timeout = ibd_timeout;
	call.attrid = IB_MLX_IS4_COUNTER_GROUP_INFO;
	call.mod = port;

	/* Counter Group Info */
	memset(&buf, 0, sizeof(buf));
	if (!ib_vendor_call_via(&buf, portid, &call, srcport))
		IBERROR("counter group info query");

	cg_info = (is4_counter_group_info_t *) & buf;
	num_cg = cg_info->num_of_counter_groups;
	printf("counter_group_info:\n");
	printf("%d counter groups\n", num_cg);
	for (i = 0; i < num_cg; i++)
		printf("group%d mask %#x\n", i, ntohl(cg_info->group_masks[i]));
}

/* Group0 counter config values */
#define IS4_G0_PortXmtDataSL_0_7  0
#define IS4_G0_PortXmtDataSL_8_15 1
#define IS4_G0_PortRcvDataSL_0_7  2

/* Group1 counter config values */
#define IS4_G1_PortXmtDataSL_8_15 1
#define IS4_G1_PortRcvDataSL_0_7  2
#define IS4_G1_PortRcvDataSL_8_15 8

static int cg0, cg1;

void config_counter_groups(ib_portid_t * portid, int port)
{
	char buf[1024];
	ib_vendor_call_t call;
	is4_config_counter_groups_t *cg_config;

	memset(&call, 0, sizeof(call));
	call.mgmt_class = IB_MLX_VENDOR_CLASS;
	call.attrid = IB_MLX_IS4_CONFIG_COUNTER_GROUP;
	call.timeout = ibd_timeout;
	call.mod = port;
	/* configure counter groups for groups 0 and 1 */
	call.method = IB_MAD_METHOD_SET;

	memset(&buf, 0, sizeof(buf));
	cg_config = (is4_config_counter_groups_t *) & buf;

	printf("counter_groups_config: configuring group0 %d group1 %d\n", cg0,
	       cg1);
	cg_config->group_selects[0].group_select = (uint8_t) cg0;
	cg_config->group_selects[1].group_select = (uint8_t) cg1;

	if (!ib_vendor_call_via(&buf, portid, &call, srcport))
		IBERROR("config counter group set");

	/* get config counter groups */
	memset(&buf, 0, sizeof(buf));
	call.method = IB_MAD_METHOD_GET;

	if (!ib_vendor_call_via(&buf, portid, &call, srcport))
		IBERROR("config counter group query");
}

static int general_info, xmit_wait, counter_group_info, config_counter_group;

static int process_opt(void *context, int ch, char *optarg)
{
	int ret;
	switch (ch) {
	case 'N':
		general_info = 1;
		break;
	case 'w':
		xmit_wait = 1;
		break;
	case 'i':
		counter_group_info = 1;
		break;
	case 'c':
		config_counter_group = 1;
		ret = sscanf(optarg, "%d,%d", &cg0, &cg1);
		if (ret != 2)
			return -1;
		break;
	default:
		return -1;
	}
	return 0;
}

int main(int argc, char **argv)
{
	int mgmt_classes[4] = { IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS,
		IB_MLX_VENDOR_CLASS
	};
	ib_portid_t portid = { 0 };
	int port = 0;
	char buf[1024];
	ib_vendor_call_t call;
	is3_general_info_t *gi;
	is3_config_space_t *cs;
	int i;

	const struct ibdiag_opt opts[] = {
		{"N", 'N', 0, NULL, "show IS3 or IS4 general information"},
		{"w", 'w', 0, NULL, "show IS3 port xmit wait counters"},
		{"i", 'i', 0, NULL, "show IS4 counter group info"},
		{"c", 'c', 1, "<num,num>", "configure IS4 counter groups"},
		{0}
	};

	char usage_args[] = "<lid|guid> [port]";
	const char *usage_examples[] = {
		"-N 6\t\t# read IS3 or IS4 general information",
		"-w 6\t\t# read IS3 port xmit wait counters",
		"-i 6 12\t# read IS4 port 12 counter group info",
		"-c 0,1 6 12\t# configure IS4 port 12 counter groups for PortXmitDataSL",
		"-c 2,8 6 12\t# configure IS4 port 12 counter groups for PortRcvDataSL",
		NULL
	};

	ibdiag_process_opts(argc, argv, NULL, "D", opts, process_opt,
			    usage_args, usage_examples);

	argc -= optind;
	argv += optind;

	if (argc > 1)
		port = strtoul(argv[1], 0, 0);

	srcport = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 4);
	if (!srcport)
		IBERROR("Failed to open '%s' port '%d'", ibd_ca, ibd_ca_port);

	if (argc) {
		if (ib_resolve_portid_str_via(&portid, argv[0], ibd_dest_type,
					      ibd_sm_id, srcport) < 0)
			IBERROR("can't resolve destination port %s", argv[0]);
	} else {
		if (ib_resolve_self_via(&portid, &port, 0, srcport) < 0)
			IBERROR("can't resolve self port %s", argv[0]);
	}

	if (counter_group_info) {
		counter_groups_info(&portid, port);
		exit(0);
	}

	if (config_counter_group) {
		config_counter_groups(&portid, port);
		exit(0);
	}

	/* These are Mellanox specific vendor MADs */
	/* but vendors change the VendorId so how know for sure ? */
	/* Only General Info and Port Xmit Wait Counters */
	/* queries are currently supported */
	if (!general_info && !xmit_wait)
		IBERROR("at least one of -N and -w must be specified");

	/* Would need a list of these and it might not be complete */
	/* so for right now, punt on this */

	memset(&call, 0, sizeof(call));
	call.mgmt_class = IB_MLX_VENDOR_CLASS;
	call.method = IB_MAD_METHOD_GET;
	call.timeout = ibd_timeout;

	memset(&buf, 0, sizeof(buf));
	/* vendor ClassPortInfo is required attribute if class supported */
	call.attrid = CLASS_PORT_INFO;
	if (!ib_vendor_call_via(&buf, &portid, &call, srcport))
		IBERROR("classportinfo query");

	memset(&buf, 0, sizeof(buf));
	call.attrid = IB_MLX_IS3_GENERAL_INFO;
	if (!ib_vendor_call_via(&buf, &portid, &call, srcport))
		IBERROR("vendstat");
	gi = (is3_general_info_t *) & buf;

	if (general_info) {
		/* dump IS3 or IS4 general info here */
		printf("hw_dev_rev:  0x%04x\n", ntohs(gi->hw_info.hw_revision));
		printf("hw_dev_id:   0x%04x\n", ntohs(gi->hw_info.device_id));
		printf("hw_uptime:   0x%08x\n", ntohl(gi->hw_info.uptime));
		printf("fw_version:  %02d.%02d.%02d\n",
		       gi->fw_info.major, gi->fw_info.minor,
		       gi->fw_info.sub_minor);
		printf("fw_build_id: 0x%04x\n", ntohl(gi->fw_info.build_id));
		printf("fw_date:     %02d/%02d/%04x\n",
		       gi->fw_info.month, gi->fw_info.day,
		       ntohs(gi->fw_info.year));
		printf("fw_psid:     '%s'\n", gi->fw_info.psid);
		printf("fw_ini_ver:  %d\n",
		       ntohl(gi->fw_info.ini_file_version));
		printf("sw_version:  %02d.%02d.%02d\n", gi->sw_info.major,
		       gi->sw_info.minor, gi->sw_info.sub_minor);
	}

	if (xmit_wait) {
		if (ntohs(gi->hw_info.device_id) != IS3_DEVICE_ID)
			IBERROR("Unsupported device ID 0x%x",
				ntohs(gi->hw_info.device_id));

		memset(&buf, 0, sizeof(buf));
		call.attrid = IB_MLX_IS3_CONFIG_SPACE_ACCESS;
		/* Limit of 18 accesses per MAD ? */
		call.mod = 2 << 22 | 16 << 16;	/* 16 records */
		/* Set record addresses for each port */
		cs = (is3_config_space_t *) & buf;
		for (i = 0; i < 16; i++)
			cs->record[i].address =
			    htonl(IB_MLX_IS3_PORT_XMIT_WAIT + ((i + 1) << 12));
		if (!ib_vendor_call_via(&buf, &portid, &call, srcport))
			IBERROR("vendstat");

		for (i = 0; i < 16; i++)
			if (cs->record[i].data)	/* PortXmitWait is 32 bit counter */
				printf("Port %d: PortXmitWait 0x%x\n", i + 4, ntohl(cs->record[i].data));	/* port 4 is first port */

		/* Last 8 ports is another query */
		memset(&buf, 0, sizeof(buf));
		call.attrid = IB_MLX_IS3_CONFIG_SPACE_ACCESS;
		call.mod = 2 << 22 | 8 << 16;	/* 8 records */
		/* Set record addresses for each port */
		cs = (is3_config_space_t *) & buf;
		for (i = 0; i < 8; i++)
			cs->record[i].address =
			    htonl(IB_MLX_IS3_PORT_XMIT_WAIT + ((i + 17) << 12));
		if (!ib_vendor_call_via(&buf, &portid, &call, srcport))
			IBERROR("vendstat");

		for (i = 0; i < 8; i++)
			if (cs->record[i].data)	/* PortXmitWait is 32 bit counter */
				printf("Port %d: PortXmitWait 0x%x\n",
				       i < 4 ? i + 21 : i - 3,
				       ntohl(cs->record[i].data));
	}

	mad_rpc_close_port(srcport);
	exit(0);
}
